File: apps/ios/Crumb/Networking/MediaUrls.swift
// SPDX-License-Identifier: AGPL-3.0-or-later

import Foundation

struct MediaUrls {
    let serverUrl: String
    let token: String?
    /// Per-camera scoped media-token cache (P0-SESSIONS), owned by
    /// `AppContainer` and shared by every `MediaUrls` value it hands out.
    /// `Optional` only so a bare `MediaUrls` can still be constructed (e.g. a
    /// future preview/test) without a live container — `scopedURL` simply
    /// returns `nil` in that case, matching "token mint failed".
    let tokenCache: MediaTokenCache?

    /// Full-JWT-authed URL — for endpoints that are NOT single-camera-scoped
    /// (export batch downloads, which can span multiple cameras and the
    /// archive pseudo-camera) or non-media API calls. Do NOT use this for
    /// per-camera media (frame/segment/clip/filmstrip) — use `scopedURL`.
    func authed(_ pathOrUrl: String) -> URL? {
        let absolute = toAbsolute(pathOrUrl)
        guard var components = URLComponents(string: absolute) else { return URL(string: absolute) }
        if let token, !token.isEmpty {
            var items = components.queryItems ?? []
            items.append(URLQueryItem(name: "token", value: token))
            components.queryItems = items
        }
        return components.url
    }

    /// Camera-scoped media URL carrying a short-lived (~15 min) `?token=` minted
    /// via `GET /media-token?camera=<cameraId>` (cached/refreshed/deduped by
    /// `MediaTokenCache`), instead of the full login JWT. This is the
    /// migrated replacement for `authed(_:)` on every per-camera media
    /// endpoint (live stream proxy, recorded segment, filmstrip/scrub frame,
    /// clip thumbnail/video, camera snapshot).
    ///
    /// Returns `nil` if the token mint fails (offline, camera access revoked,
    /// session expired — the 401 path already routes through
    /// `CrumbAPI.execute`'s `clearSession()`) — callers must NOT fall back to
    /// the full JWT; let the existing player/image retry loop try again.
    func scopedURL(cameraId: String, _ pathOrUrl: String) async -> URL? {
        guard let tokenCache, let token = await tokenCache.token(for: cameraId) else { return nil }
        return urlWithToken(pathOrUrl, token: token)
    }

    /// Non-async convenience for call sites that already hold a fresh token
    /// (e.g. re-minting a per-segment URL from a token obtained moments
    /// earlier in the same async context — see `HEVCRetag`'s range-proxy).
    func urlWithToken(_ pathOrUrl: String, token: String) -> URL? {
        let absolute = toAbsolute(pathOrUrl)
        guard var components = URLComponents(string: absolute) else { return URL(string: absolute) }
        var items = components.queryItems ?? []
        items.append(URLQueryItem(name: "token", value: token))
        components.queryItems = items
        return components.url
    }

    func cameraFrameUrl(_ cameraId: String) async -> URL? {
        await scopedURL(cameraId: cameraId, "/cameras/\(cameraId)/frame.jpg")
    }

    /// Authenticated live fragmented-MP4 stream via the API's `/live` proxy
    /// (`GET /live/{id}/stream.mp4?stream=main|sub`). go2rtc is no longer exposed
    /// directly (it runs with `local_auth`), so live rides the JWT-protected API
    /// with a scoped `?token=` like every other media URL. Mint fresh per connect:
    /// the token is short-lived and a persistent stream may reconnect after it
    /// expires.
    func liveFmp4URL(cameraId: String, sub: Bool) async -> URL? {
        await scopedURL(cameraId: cameraId, "/live/\(cameraId)/stream.mp4?stream=\(sub ? "sub" : "main")")
    }

    /// Authenticated WebRTC (WHEP) signaling endpoint via the API's `/live` proxy
    /// (`POST /live/{id}/webrtc?stream=main|sub`) — go2rtc's REST API is no longer
    /// LAN-reachable. Scoped `?token=`, minted fresh per signaling POST (the
    /// manager reconnects, and the token is short-lived).
    func liveWhepURL(cameraId: String, sub: Bool) async -> URL? {
        await scopedURL(cameraId: cameraId, "/live/\(cameraId)/webrtc?stream=\(sub ? "sub" : "main")")
    }

    /// Historical still extracted on-demand from recorded footage at `ts`
    /// (RFC-3339). Used for the playback wall's scrub-to-moment tile previews.
    func historicalFrameUrl(cameraId: String, tsISO: String) async -> URL? {
        let encoded = tsISO.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? tsISO
        return await scopedURL(cameraId: cameraId, "/filmstrip/\(cameraId)/frame?ts=\(encoded)")
    }

    /// Thumbnail still for a clip. Requires token auth; scoped to `cameraId`.
    func clipThumbUrl(_ id: String, cameraId: String) async -> URL? {
        await scopedURL(cameraId: cameraId, "/clip/\(id)/thumbnail.jpg")
    }

    /// Preview MP4 for a clip. `quality` defaults to `"preview"` (reduced res/fps).
    func clipVideoUrl(_ id: String, cameraId: String, quality: String = "preview") async -> URL? {
        await scopedURL(cameraId: cameraId, "/clip/\(id)/clip.mp4?q=\(quality)")
    }


    private func toAbsolute(_ pathOrUrl: String) -> String {
        if pathOrUrl.hasPrefix("http://") || pathOrUrl.hasPrefix("https://") {
            return pathOrUrl
        }
        let base = serverUrl.hasSuffix("/") ? String(serverUrl.dropLast()) : serverUrl
        let path = pathOrUrl.hasPrefix("/") ? pathOrUrl : "/\(pathOrUrl)"
        return "\(base)\(path)"
    }
}

File: apps/android/app/src/main/java/video/crumb/app/data/MediaUrls.kt
// SPDX-License-Identifier: AGPL-3.0-or-later

package video.crumb.app.data

import android.net.Uri

/**
 * Builds absolute media URLs.
 *
 * ExoPlayer (segment playback), Coil (still frames, filmstrip/clip thumbnails)
 * fetch bytes directly and cannot set an Authorization header, so the API
 * accepts a token on the query string for media endpoints. Two flavours:
 *
 * - **Scoped, per-camera** (segments, filmstrip frames, camera stills, clip
 *   thumbnails/video): built by the `scoped*` / `authedScoped` suspend
 *   functions below, which use [MediaTokenCache] to attach a short-lived
 *   (~15 min) token valid ONLY for that one camera — never the full login JWT.
 * - **Cross-camera / archive** (export downloads): still carries the full
 *   login JWT via [authed], since an export can span multiple cameras and
 *   stages that a single-camera scoped token can't authorize. See
 *   `ExportViewModel.authedUrl`.
 *
 * RTSP live URLs come pre-formed from the API (go2rtc) and are returned as-is;
 * go2rtc on the LAN does not require a token at all.
 */
class MediaUrls(
    private val serverUrl: String,
    private val mediaTokenCache: MediaTokenCache,
) {
    /**
     * Resolve a possibly-relative API path against the server base, adding the
     * per-camera scoped token as `?token=`. Fetches/refreshes the cached token
     * for [cameraId] first (suspends only on a cache miss / near-expiry).
     */
    private suspend fun authedScoped(cameraId: String, pathOrUrl: String): String {
        val absolute = toAbsolute(pathOrUrl)
        val token = mediaTokenCache.freshToken(cameraId)
        val sep = if (absolute.contains('?')) '&' else '?'
        return "$absolute${sep}token=${Uri.encode(token)}"
    }

    /** RTSP/WebRTC URLs from the API are already absolute and unauthenticated. */
    fun raw(url: String): String = url

    /**
     * API-proxied still-frame URL for the playback wall and any consumer that needs
     * a preview frame regardless of which go2rtc instance owns the camera.
     *
     * Routes to `GET /cameras/{id}/frame.jpg` (scoped-token authed), which the
     * backend proxies from whichever go2rtc actually has the camera. Use this for all
     * wall preview tiles — it works for cameras migrated to Crumb's go2rtc and for
     * cameras that are still on Frigate's go2rtc.
     */
    suspend fun cameraFrameUrl(cameraId: String): String =
        authedScoped(cameraId, "/cameras/$cameraId/frame.jpg")

    /** Scoped-token clip thumbnail (Coil) for a Clips-feed [clipId] on [cameraId]. */
    suspend fun clipThumbUrl(cameraId: String, clipId: String): String =
        authedScoped(cameraId, "/clip/${Uri.encode(clipId)}/thumbnail.jpg")

    /**
     * Scoped-token clip mp4 (ExoPlayer) for a Clips-feed [clipId] on [cameraId].
     *
     * [quality] is `"preview"` (small, reduced res/fps — the feed default) or
     * `"full"` (source resolution, generated on demand for an explicit
     * full-quality action).
     */
    suspend fun clipVideoUrl(cameraId: String, clipId: String, quality: String = "preview"): String =
        authedScoped(cameraId, "/clip/${Uri.encode(clipId)}/clip.mp4?q=$quality")

    /**
     * Scoped-token URL for any other [cameraId]-owned relative media path/URL
     * (recorded segment files, filmstrip frame URLs — both returned by the API
     * as camera-scoped relative paths).
     */
    suspend fun scopedUrl(cameraId: String, pathOrUrl: String): String = authedScoped(cameraId, pathOrUrl)

    /**
     * Resolve [pathOrUrl] against the server base, adding the full LOGIN JWT as
     * `?token=`. Reserved for media that isn't scoped to one camera — currently
     * only export downloads (`ExportViewModel.authedUrl`), which can span
     * multiple cameras/archive stages. Do NOT use this for any new per-camera
     * media URL — use [scopedUrl] / [cameraFrameUrl] / [clipThumbUrl] /
     * [clipVideoUrl] instead so per-camera media never carries the full JWT.
     */
    fun authed(pathOrUrl: String, loginToken: String?): String {
        val absolute = toAbsolute(pathOrUrl)
        if (loginToken.isNullOrBlank()) return absolute
        val sep = if (absolute.contains('?')) '&' else '?'
        return "$absolute${sep}token=${Uri.encode(loginToken)}"
    }

    private fun toAbsolute(pathOrUrl: String): String {
        if (pathOrUrl.startsWith("http://") || pathOrUrl.startsWith("https://")) return pathOrUrl
        val base = serverUrl.trimEnd('/')
        val path = if (pathOrUrl.startsWith("/")) pathOrUrl else "/$pathOrUrl"
        return "$base$path"
    }
}
